source/opt/struct_packing_pass.cpp

```diff
@@ -0,0 +1,463 @@
+  if (s == "hlslCbuffer") return PackingRules::HlslCbuffer;
+  if (s == "hlslCbufferPackOffset") return PackingRules::HlslCbufferPackOffset;
+  if (s == "scalar") return PackingRules::Scalar;
+  if (s == "scalarEnhancedLayout") return PackingRules::ScalarEnhancedLayout;
+  return PackingRules::Undefined;
+}
+
+StructPackingPass::StructPackingPass(const char* structToPack,
+                                     PackingRules rules)
+    : structToPack_{structToPack != nullptr ? structToPack : ""},
+      packingRules_{rules} {}
+
+Pass::Status StructPackingPass::Process() {
+  if (packingRules_ == PackingRules::Undefined) return Status::Failure;
+
+  buildConstantsMap();
+
+  // Find all structs of interest
+  uint32_t structIdToPack = ~0u;
+  for (Instruction& instr : context()->module()->debugs2()) {
+    if (instr.opcode() == spv::Op::OpName &&
+        instr.GetOperand(1).AsString() == structToPack_) {
+      structIdToPack = instr.GetOperand(0).AsId();
+      break;
+    }
+  }
+
+  struct StructMemberInfo {
+    spv::Id typeId = 0;
+    analysis::Type* type = nullptr;
+    uint32_t offset = 0; // Original offset
+    uint32_t packedOffset = 0; // Packed offset
+  };
+  std::vector<StructMemberInfo> structMembers;
+
+  if (structIdToPack == ~0u) {
+    return Status::SuccessWithoutChange;
+  }
+
+  bool modified = false;
+
+  // Find all struct member types
+  for (auto
+           it =
+               std::reverse_iterator{context()->module()->types_values_end()},
+           itEnd =
+               std::reverse_iterator{context()->module()->types_values_begin()};
+       it != itEnd; ++it) {
+    if (it->opcode() == spv::Op::OpTypeStruct &&
+        it->result_id() == structIdToPack) {
+      // Found struct type to pack, now collect all types of its members
+      assert(it->NumOperands() > 0 &&
+             "Number of operands in OpTypeStruct instruction must not be zero");
+      const uint32_t numMembers = it->NumOperands() - 1;
+      structMembers.resize(numMembers);
+      for (uint32_t i = 0; i < numMembers; ++i) {
+        const spv::Id memberTypeId = it->GetOperand(1 + i).AsId();
+        if (analysis::Type* memberType =
+                context()->get_type_mgr()->GetType(memberTypeId)) {
+          structMembers[i].typeId = memberTypeId;
+          structMembers[i].type = memberType;
+        }
+      }
+      break;
+    }
+  }
+
+  // Returns true if the specified instruction is a OpMemberDecorate for the
+  // struct we're looking for with an offset decoration
+  auto isMemberOffsetDecoration = [structIdToPack](Instruction& instr) -> bool {
+    return instr.opcode() == spv::Op::OpMemberDecorate &&
+           instr.GetOperand(0).AsId() == structIdToPack &&
+           static_cast<spv::Decoration>(instr.GetOperand(2).words[0]) ==
+               spv::Decoration::Offset;
+  };
+
+  // Find and re-assign all member offset decorations
+  for (auto it = context()->module()->annotation_begin(),
```

Comment:
I would recommend using `WhileEachDecoration` from the decoration manager.

Reply:
It seems all `While`/`ForEachDecoration` functions only provide a constant instruction, so this makes modifying the decoration a bit tricky. It would also require to use extra fields to track the state of the next offset instead of using local variables. I think these iteration functions work better when operating on each decoration independently, but in this case, subsequent instructions are depending on the immediate previous one.